### PR TITLE
[CI CD] added direct maven central publish

### PR DESCRIPTION
**Pull request by YannickWeber**
* adapted build.gradle.kts for direct maven central publish
* added GitHub action that performs the upload automatically upon release creation

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -117,46 +118,43 @@ tasks.javadoc {
...
+                }
+                developers {
+                    project.metadata.developers
+                }
+                scm {
+                    connection.set("scm:git:git://github.com/hivemq/${project.metadata.github!!.repo}.git")
+                    developerConnection.set("scm:git:ssh://github.com/hivemq/${project.metadata.github!!.repo}.git")
+                    url.set("http://github.com/hivemq/${project.metadata.github!!.repo}/")
+                }
+            }
```
Comment (SgtSilvio) [inline build.gradle.kts:138]:
```suggestion
            from(components["java"])
```
all this is done by the metadata plugin to be compatible with maven central out of the box (which is one reason that this gradle plugin exists)
Also, the `from(components["java"])` was deleted

File: build.gradle.kts [hunk 2]
```diff
@@ -3,11 +3,12 @@ import java.util.*
 plugins {
     id("java-library")
     id("maven-publish")
-    id("com.jfrog.bintray")
     id("com.github.hierynomus.license")
     id("com.github.sgtsilvio.gradle.utf8")
     id("com.github.sgtsilvio.gradle.metadata")
     id("com.github.sgtsilvio.gradle.javadoc-links")
+    id("io.github.gradle-nexus.publish-plugin")
+    signing
```
Comment (SgtSilvio) [inline build.gradle.kts:11]:
```suggestion
    id ("signing")
```
(I like consistency)

File: build.gradle.kts [hunk 3]
```diff
@@ -117,46 +118,43 @@ tasks.javadoc {
...
-                sync = false
-                user = "${project.findProperty("mavenCentralUser") ?: System.getenv("MAVEN_CENTRAL_USER")}"
-                password = "${project.findProperty("mavenCentralKey") ?: System.getenv("MAVEN_CENTRAL_KEY")}"
-                close = "0"
-            }
+nexusPublishing {
+    repositories {
+        sonatype {
+            username.set("${project.findProperty("sonatypeUser") ?: System.getenv("SONATYPE_USERNAME")}")
+            password.set("${project.findProperty("sonatypePassword") ?: System.getenv("SONATYPE_PASSWORD")}")
```
Comment (SgtSilvio) [inline build.gradle.kts:147]: should be unnecessary if you use the convention from the plugin
> you need to set the sonatypeUsername and sonatypePassword project properties, e.g. in ~/.gradle/gradle.properties or via the ORG_GRADLE_PROJECT_sonatypeUsername and ORG_GRADLE_PROJECT_sonatypePassword environment variables.

File: gradle.properties [hunk 4]
```diff
@@ -12,3 +12,4 @@ plugin.bintray.version=1.8.5
 plugin.utf8.version=0.1.0
 plugin.metadata.version=0.2.0
 plugin.javadoc-links.version=0.3.0
+plugin.io.github.gradle-nexus.publish-plugin.version=1.0.0
```
Comment (SgtSilvio) [inline gradle.properties:15]:
```suggestion
plugin.nexus-publish.version=1.0.0
```

File: build.gradle.kts (hunk 3 above)
Comment (YannickWeber) [reply build.gradle.kts:147]: Yes, but I want to use organisation wide Environment variables and want to keep it consistent over all projects.

File: settings.gradle.kts [hunk 5]
```diff
@@ -7,5 +7,6 @@ pluginManagement {
         id("com.github.sgtsilvio.gradle.utf8") version "${extra["plugin.utf8.version"]}"
         id("com.github.sgtsilvio.gradle.metadata") version "${extra["plugin.metadata.version"]}"
         id("com.github.sgtsilvio.gradle.javadoc-links") version "${extra["plugin.javadoc-links.version"]}"
+        id("io.github.gradle-nexus.publish-plugin") version "${extra["plugin.io.github.gradle-nexus.publish-plugin.version"]}"
```
Comment (SgtSilvio) [inline settings.gradle.kts:10]:
```suggestion
        id("io.github.gradle-nexus.publish-plugin") version "${extra["plugin.nexus-publish.version"]}"
```
(I like to keep it readable)

File: build.gradle.kts (hunk 3 above)
Comment (SgtSilvio) [reply build.gradle.kts:147]: Of course you can name the env vars in GitHub/Jenkins whatever you want, just pass them to gradle in the right format

Comment (SgtSilvio) [reply build.gradle.kts:147]: Sticking to conventions is really important for build maintenance imho

Comment (SgtSilvio) [reply build.gradle.kts:147]: for example, you can assign different env names in the github workflow, or pass it to the gradle command like `ORG_GRADLE_PROJECT_sonatypeUsername=$SONATYPE_USERNAME ./gradlew ...` (I hope this is the correct syntax)

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+          java-version: "11"
+      - name: Check
+        run: ./graldew check
+      - name: Publish to Maven Central
+        env:
+          SIGN_KEY: ${{ secrets.SIGN_KEY }}
+          SIGN_KEY_PASS: ${{ secrets.SIGN_KEY_PASS }}
+          SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
+          SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
+        run: ./gradlew clean publishToSonatype closeAndReleaseSonatypeStagingRepository
```
Comment (SgtSilvio) [inline .github/workflows/publish.yml:26]:
```suggestion
        run: ./gradlew publishToSonatype closeAndReleaseSonatypeStagingRepository
```

File: build.gradle.kts (hunk 3 above)
Comment (YannickWeber) [reply build.gradle.kts:147]: Oh, I finally get your comment now! I thought the `ORG_GRADLE_PROJECT` is a specific prefix that has to change in every project. 

`ORG_GRADLE_PROJECT_something` makes gradle set the project property `something`. It has actually nothing to do with the plugin. 

🤦🏿

**Review by DC2-DanielKrueger — APPROVED** [review]

**Pull request merged**